x/oracle/keeper/update_exchange_rates.go

```diff
@@ -84,11 +85,39 @@ func (k Keeper) getPairBallotsMapAndWhitelistedPairs(
 }
 
 // resetExchangeRates removes all exchange rates from the state
-func (k Keeper) resetExchangeRates(ctx sdk.Context) {
+// We remove the price for pair with expired prices or valid ballots
+func (k Keeper) resetExchangeRates(ctx sdk.Context, pairBallotsMap map[asset.Pair]types.ExchangeRateBallots) {
+	params, _ := k.Params.Get(ctx)
+	expirationBlocks := params.ExpirationBlocks
+
+	for _, key := range k.ExchangeRates.Iterate(ctx, collections.Range[asset.Pair]{}).Keys() {
+		_, validBallot := pairBallotsMap[key]
+		exchangeRate, _ := k.ExchangeRates.Get(ctx, key)
+		isExpired := exchangeRate.BlockSet+expirationBlocks <= uint64(ctx.BlockHeight())
+
+		if validBallot || isExpired {
+			err := k.ExchangeRates.Delete(ctx, key)
+			if err != nil {
+				panic(err)
+			}
+		}
+	}
+}
+
+// CleanExpiredRates removes all exchange rates from the state that are expired
+func (k Keeper) CleanExpiredRates(ctx sdk.Context) {
```

Comment:
Technically won't `resetExchangeRates` get rid of expired prices for us? So we don't need to call `CleanExpiredRates` at the end of every block, and we can re-use the iteration over the ballots at the end of every vote period instead (more efficient for computation).

Reply:
If we set expiration blocks as a multiple of voting period, then blocks will expire at the end of a voting period anyways so checking every block wouldn't be necessary. I think it would be worthwhile to save the compute, otherwise we could be slowing down our block times with a lot of state reads in EndBlocker.